TriangleNet_Engine/Compute/VoronoiRegions.cs

```diff
@@ -0,0 +1,313 @@
+        [Input("openingCurves", "Inner openings to be cut out from the voronoi cells. Must be coplanar with the provided points.")]
+        [Input("plane", "Optional plane for the voronoi. If provided, all points must be complanar with the plane. If nothing provided, a best fit plane will be calculated. For colinear points, if nothing no plane provided, a plane aligned with the global Z-axis will be created.")]
+        [Input("boundarySize", "To handle problems at boundaries, extra points are added outside the bounds of the provided points for the generation of the voronoi and then culled away. This value controls how far off these points should be created. If a negative value is provided, this value will be calculated automatically, based on the size of the boundingbox of the provided points.", typeof(Length))]
+        [Input("tolerance", "Tolerance to be used in the method.", typeof(Length))]
+        [Output("regions", "Voronoi regions calculated by the method. The position in the list will correspond to the position in the list of the provided points.")]
+        public static List<List<PolyCurve>> VoronoiRegions(List<Point> points, ICurve boundaryCurve, List<ICurve> openingCurves = null, Plane plane = null, double boundarySize = -1, double tolerance = Tolerance.Distance)
+        {
+            openingCurves = openingCurves ?? new List<ICurve>();
+
+            List<Point> checkingPoints = new List<Point>(points);
+            checkingPoints.AddRange(boundaryCurve.IControlPoints());
+            checkingPoints.AddRange(openingCurves.SelectMany(x => x.IControlPoints()));
+
+            if (!checkingPoints.IsCoplanar(tolerance))
+            {
+                Engine.Reflection.Compute.RecordError("The points, boundaryCurve and openingCurves all need to be coplanar.");
+                return new List<List<PolyCurve>>();
+            }
+
+            List<Polyline> untrimmedRegions = VoronoiRegions(points, plane, boundarySize, tolerance);
+
+            List<List<PolyCurve>> boundaryTrimmedCurves = untrimmedRegions.Select(x => x.BooleanIntersection(boundaryCurve, tolerance)).ToList();
+
+            List<List<PolyCurve>> trimmedCurves = new List<List<PolyCurve>>();
+
+            foreach (List<PolyCurve> pointCurves in boundaryTrimmedCurves)
+            {
+                List<PolyCurve> curves = new List<PolyCurve>();
+
+                foreach (PolyCurve curve in pointCurves)
+                {
+                    curves.AddRange(curve.BooleanDifference(openingCurves, tolerance));
+                }
+                trimmedCurves.Add(curves);
+            }
+
+            return trimmedCurves;
+        }
+
+        /***************************************************/
+        /****      Private Methods                       ****/
+        /***************************************************/
+
+        [Description("Creates a set of voronoi cells from a list of colinear points. Requires the incoming points to be colinear to work.")]
```

Comment:
Why does this method not protect against bad input (i.e. non-colinear points)? Why is the onus on the user calling the method (user being developer in this instance) to remember this and risk bad input?

Reply:
Decided against it as the method is private and a check already has been made for the use case to avoid checking the same thing twice. This is basically a fallback/edge case method due to the fact that neither fit plane or triangulation works for colinear points. Started writing this in the main method, but as it started to grow I thought better put it in a private method instead.

Reply:
Yeah agree with it being private. Just not sure I like that the responsibility of checking if this method can work as desired is resting on the calling methods rather on this method. Ideally the method should protect itself, but given the work here is small and will likely only be called by the methods you've written I'll accept not doing anything if you wish not to.

Reply:
Just want to avoid running a check that has already been run again.

Reply:
Yeah I see where you're coming from. My main concern is say someone like @alelom comes along to add another method and uses your method but doesn't realise his method should be doing the check, it might be run wrongly. I do see why, in the way you've developed the code for this PR you wouldn't want to put the check in again. As I say above, am happy to accept without given the scope and nature, but I think it's worth considering - but will accept no change if your consideration is not to :smile: